slackbotExercise.py

```diff
@@ -3,90 +3,280 @@
-    URLTOKENSTRING = settings["URLTOKENSTRING"]
-    TEAMNAMESTRING = settings["TEAMNAMESTRING"]
+    # The max number of users we are willing to look forward
+    # to try and find a good match
+    sliding_window = bot.sliding_window_size
 
+    # find a user to draw, priority going to first in
+    for i in range(len(bot.user_queue)):
+        user = bot.user_queue[i]
 
-# Extracts online users from Slack API
-def extractSlackUsers(token):
-    # Set token parameter of Slack API call
-    tokenString = token
-    params = {"token": tokenString}
+        # User should be active and not have done exercise yet
+        if user in active_users and not user.hasDoneExercise(exercise):
+            bot.user_queue.remove(user)
+            return user
+        elif user in active_users:
+            # Decrease sliding window by one. Basically, we don't want to jump
+            # too far ahead in our queue
+            sliding_window -= 1
+            if sliding_window <= 0:
+                break
 
-    # Capture Response as JSON
-    response = requests.get("https://slack.com/api/users.list", params=params)
-    users = json.loads(response.text, encoding='utf-8')["members"]
+    # If everybody has done exercises or we didn't find a person within our sliding window,
+    for user in bot.user_queue:
+        if user in active_users:
+            bot.user_queue.remove(user)
+            return user
 
-    def findUserNames(x):
-        if getStats(x) is False:
-            return None
-        name = "@" + x["name"].encode('utf-8')
-        return name.encode('utf-8')
+    # If we weren't able to select one, just pick a random
+    print "Selecting user at random (queue length was " + str(len(bot.user_queue)) + ")"
+    return active_users[random.randrange(0, len(active_users))]
 
-    def getStats(x):
-        params = {"token": tokenString, "user": x["id"]}
-        response = requests.get("https://slack.com/api/users.getPresence", params=params)
-        status = json.loads(response.text, encoding='utf-8')["presence"]
-        return status == "active"
 
-    return filter(None, list(map(findUserNames, users)))
+'''
+Fetches a list of all active users in the channel
+'''
+def fetchActiveUsers(bot):
+    # Check for new members
+    params = {"token": USER_TOKEN_STRING, "channel": bot.channel_id}
+    response = requests.get("https://slack.com/api/channels.info", params=params)
+    user_ids = json.loads(response.text, encoding='utf-8')["channel"]["members"]
 
+    active_users = []
 
-# Selects Next Time Interval and Returns the Exercise
-def selectExerciseAndStartTime():
+    for user_id in user_ids:
+        # Add user to the cache if not already
+        if user_id not in bot.user_cache:
+            bot.user_cache[user_id] = User(user_id)
+            if not bot.first_run:
+                # Push our new users near the front of the queue!
+                bot.user_queue.insert(2,bot.user_cache[user_id])
```

Comment:
```suggestion
        if user_id in bot.user_cache:
            continue
        bot.user_cache[user_id] = User(user_id)
        if bot.first_run:
            continue
        # Push our new users near the front of the queue!
        bot.user_queue.insert(2,bot.user_cache[user_id])
```
this looks like less branches
(I'm just trying to do my homework, sorry)